### PR TITLE
Enable distributed CFG for SD3 pipeline

### PR DESCRIPTION
**Pull request by dsocek**
# What does this PR do?

🚀 Adds distributed CFG inference support for SD3 models and updates documentation
⚡ Replaces standard RMSNorm with Gaudi-optimized RMSNorm in SD3 attention, improving performance
🛠️ Fixes text-to-image sample script to correctly output generated images during multi-card inference

# Distributed CFG Inference

In Optimum-Habana Diffusers, multi-card inference currently spawns independent image generations on each worker, which does not improve performance per device. However, the SD3 pipeline architecture allows for a performance optimization: by leveraging classifier-free guidance (CFG) mode, we can parallelize the conditional and unconditional branches across two devices, yielding up to ~2× speedup per image. This approach is not yet supported in the main Hugging Face Diffusers library (for general GPU usage), making this an original implementation.

**SD3 Inference**

In SD3x family of models Diffusion Transformer (DiT) is used in the denoising loop for image generation.  When we have prompt (and optionally negative prompt) and guidance scale > 1 (default recommended setting is 7.5), then SD3 pipeline runs in a so-called classifier-free guidance (CFG) mode.  In this mode (which is default), we process 2 latents during denoising loop:

1. Prompt-guided latents
2. Non-guided latents (negative prompt guided if negative prompt is supplied)

This means that we need 2 forward passes, one for guided latents and the other for non-guided latents. In Diffusers, the latents and embeddings are packed into tensors with a batch of size 2. Effectively, we do not perform 2 forward passes, but we run 1 forward pass with batch size 2. This process is illustrated with the following figure:

image

**Distributed 2-Card CFG SD3 Inference**

The core idea behind distributed classifier-free guidance (CFG) inference across two devices is to parallelize the conditional and unconditional branches of the denoising process. We start by running the model inference independently on each device. However, at the denoising stage, the latent batch is split into guided and unguided components:

* Rank 0 (device 0) handles the guided (conditional) batch.
* Rank 1 (device 1) handles the unguided (unconditional) batch.

Each device performs denoising independently on its respective half of the batch. Once complete, the predicted noise tensors are exchanged between devices using an HCCL send/receive operation to reassemble the full batch and apply CFG. To ensure consistency between devices, both workers must use the same random seed during sampling, guaranteeing identical noise initialization and timestep behavior.

image

## Example

An example of performance boost and quality of generated images can be checked with the following example under `optimum-habana/examples/stable-diffusion` folder.

An example showing the performance boost and quality of generated images with distributed CFG inference mode can be ran in `optimum-habana/examples/stable-diffusion` .

Running 4 inferences for SD3.5-Large on 1xGaudi2 card:

```bash
PT_HPU_LAZY_MODE=1 python text_to_image_generation.py \
    --model_name_or_path stabilityai/stable-diffusion-3.5-large \
    --prompts "Sailing ship painting by Van Gogh" \
    --num_images_per_prompt 4 \
    --batch_size 1 \
    --num_inference_steps 28 \
    --image_save_dir /tmp/stable_diffusion_3_images_1x \
    --scheduler default \
    --use_habana \
    --use_hpu_graphs \
    --gaudi_config Habana/stable-diffusion \
    --sdp_on_bf16 \
    --bf16
```

Output:
```bash
...
[INFO|pipeline_stable_diffusion_3.py:546] 2025-06-04 22:58:26,110 >> 1 prompt(s) received, 4 generation(s) per prompt, 1 sample(s) per batch, 4 total batch(es).
100%|██████████| 28/28 [00:29<00:00,  1.05s/it]
100%|██████████| 28/28 [00:28<00:00,  1.03s/it]
100%|██████████| 28/28 [00:09<00:00,  2.90it/s]
100%|██████████| 28/28 [00:09<00:00,  2.90it/s]
[INFO|pipeline_stable_diffusion_3.py:715] 2025-06-04 22:59:58,336 >> Speed metrics: {'generation_runtime': 79.8969, 'generation_samples_per_second': 0.099, 'generation_steps_per_second': 2.768}
```
=> performance of **~2.9** steps/sec after warmup.

Generated images:
image

Running 4 inferences for SD3.5-Large on 2xGaudi2 cards in Distributed CFG mode:
```bash
PT_HPU_LAZY_MODE=1 \
python ../gaudi_spawn.py --world_size 2 text_to_image_generation.py \
    --model_name_or_path stabilityai/stable-diffusion-3.5-large \
    --prompts "Sailing ship painting by Van Gogh" \
    --num_images_per_prompt 4 \
    --batch_size 1 \
    --num_inference_steps 28 \
    --image_save_dir /tmp/stable_diffusion_3_images \
    --scheduler default \
    --use_habana \
    --use_hpu_graphs \
    --use_distributed_cfg \
    --gaudi_config Habana/stable-diffusion \
    --sdp_on_bf16 \
    --bf16
```

Output:
```bash
06/04/2025 22:41:32 - INFO - __main__ - Generating images using pipeline GaudiStableDiffusion3Pipeline with scheduler GaudiFlowMatchEulerDiscreteScheduler
[INFO|pipeline_stable_diffusion_3.py:544] 2025-06-04 22:41:32,486 >> 1 prompt(s) received, 4 generation(s) per prompt, 1 sample(s) per batch, 4 total batch(es).
06/04/2025 22:41:42 - INFO - __main__ - Generating images using pipeline GaudiStableDiffusion3Pipeline with scheduler GaudiFlowMatchEulerDiscreteScheduler
[INFO|pipeline_stable_diffusion_3.py:544] 2025-06-04 22:41:42,959 >> 1 prompt(s) received, 4 generation(s) per prompt, 1 sample(s) per batch, 4 total batch(es).
100%|██████████| 28/28 [00:29<00:00,  1.04s/it]
100%|██████████| 28/28 [00:19<00:00,  1.46it/s]
100%|██████████| 28/28 [00:18<00:00,  1.49it/s]
100%|██████████| 28/28 [00:18<00:00,  1.49it/s]
100%|██████████| 28/28 [00:04<00:00,  5.62it/s]
100%|██████████| 28/28 [00:04<00:00,  5.62it/s]
100%|██████████| 28/28 [00:04<00:00,  5.63it/s]
100%|██████████| 28/28 [00:04<00:00,  5.63it/s]
[INFO|pipeline_stable_diffusion_3.py:713] 2025-06-04 22:42:44,120 >> Speed metrics: {'generation_runtime': 49.5514, 'generation_samples_per_second': 0.191, 'generation_steps_per_second': 5.343}
[INFO|pipeline_stable_diffusion_3.py:713] 2025-06-04 22:42:44,120 >> Speed metrics: {'generation_runtime': 59.5632, 'generation_samples_per_second': 0.191, 'generation_steps_per_second': 5.341}
```
=> performance of **~5.6** steps/sec after warmup.

Generated images:
image

Note that generated images for rank 0 and rank 1 devices are identical (as expected).  They are also identical to 1 card run example since randomization seed was the same.

### CODE REVIEW

File: optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py [hunk 1]
```diff
@@ -602,18 +622,40 @@ def __call__(
...
-                    )
-
-                    # perform guidance
-                    if self.do_classifier_free_guidance:
-                        noise_pred_uncond, noise_pred_text = noise_pred.chunk(2)
-                        noise_pred = noise_pred_uncond + self.guidance_scale * (noise_pred_text - noise_pred_uncond)
+                    # noise prediction (transformer call)
+                    if use_distributed_cfg and self.do_classifier_free_guidance:
+                        idx = rank % 2
+                        noise_pred_b1 = self.transformer_hpu(
```
Comment (yafshar) [inline optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py:631]: @dsocek it sounds like that the user is allowed to use more than 2 devices (even numbers). Does it mean each batch of 2 is for one image?

**Comment by yafshar** [comment]
@dsocek thanks! Very nice. I will do some testing but it looks very good to me! Thanks

File: examples/stable-diffusion/text_to_image_generation.py [hunk 2]
```diff
@@ -693,14 +704,15 @@ def main():
 
             image_save_dir.mkdir(parents=True, exist_ok=True)
             logger.info(f"Saving images in {image_save_dir.resolve()}...")
+            rank_ext = f"_rank{os.getenv('RANK')}" if int(os.getenv("WORLD_SIZE", "1")) > 1 else ""
```
Comment (yafshar) [inline examples/stable-diffusion/text_to_image_generation.py]: @dsocek this approach technically generates the same images on both devices, why do we need to write them both? It is OK to do that for debugging or testing, but not for production. The user wants one set of images

File: optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py [hunk 3]
```diff
@@ -403,6 +404,26 @@ def __call__(
         import habana_frameworks.torch as ht
         import habana_frameworks.torch.core as htcore
 
+        # Enable distributed CFG (classifier-free guidance) across 2 devices
+        # (conditional and unconditional parts are processed separately) if specified
+        use_distributed_cfg = kwargs.get("use_distributed_cfg", None)
+        if use_distributed_cfg:
```
Comment (yafshar) [inline optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py:412]:
```suggestion
        if use_distributed_cfg is not None:
```

File: optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py [hunk 4]
```diff
@@ -403,6 +404,26 @@ def __call__(
...
+                raise ValueError(f"Error: Distributed CFG requires running with at least 2 devices")
+            if not dist.is_initialized():
+                dist.init_process_group(backend="hccl", rank=rank, world_size=world_size)
+            if dist.get_world_size() % 2 != 0:
+                raise ValueError(f"Error: Distributed CFG requires even world size, but got world_size={world_size}")
+            if guidance_scale <= 1:
+                raise ValueError(
+                    "Error: Distributed CFG requires use of classifier-free guidance (guidance scale > 1)"
+                )
+            htcore.hpu.set_device(rank)
```
Comment (yafshar) [inline optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py]: @dsocek this section should not be in the forward call?

File: optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py (hunk 3 above)
Comment (dsocek) [reply optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py:412]: I was hoping to also handle a case when user passes `use_distributed_cfg=False` here

File: examples/stable-diffusion/text_to_image_generation.py (hunk 2 above)
Comment (mkpatel3-github) [reply examples/stable-diffusion/text_to_image_generation.py]: This is a sample, so it may be good to give the user the option to verify and check that both images are identical as expected and not have this in a production pipeline?

File: optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py (hunk 3 above)
Comment (yafshar) [inline optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py:412]:
```suggestion
        if use_distributed_cfg is not None and use_distributed_cfg:
```

this is a personal choice, ignore if you do not like it

Comment (dsocek) [reply optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py:412]: I am thinking to make use_distributed_cfg actual optional function argument of boolean type with False default to make it clearer and cleaner.  Will do in next iteration

File: optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py (hunk 1 above)
Comment (dsocek) [reply optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py:631]: Actually for 1 device the SD3 (normal mode) creates batch of size 2 internally.  With distributed mode we run each part (conditional and unconditional) separately on 2 devices (to generate 1 image).  So if we have N devices, N is even, then we can generate N/2 unique images.

File: examples/stable-diffusion/text_to_image_generation.py (hunk 2 above)
Comment (dsocek) [reply examples/stable-diffusion/text_to_image_generation.py]: I updated the code so now it outputs 1 image per pair of ranks (if `use_distributed_cfg` mode is enabled).

For example, if we want 4 images with 4 cards and enable with command
```batch
PT_HPU_LAZY_MODE=1 python ../gaudi_spawn.py --world_size 4 text_to_image_generation.py     --model_name_or_path stabilityai/stable-diffusion-3.5-large     --prompts "Sailing ship painting by Van Gogh"     --num_images_per_prompt 4     --batch_size 1     --num_inference_steps 28     --image_save_dir /tmp/stable_diffusion_3_images_2x     --scheduler default     --use_habana     --use_hpu_graphs     --use_distributed_cfg     --gaudi_config Habana/stable-diffusion     --sdp_on_bf16     --bf16
```
We now see output of 8 images (2 2xHPU x 4 images per prompt) as follows:
```bash
$ ls /tmp/stable_diffusion_3_images_2x
image_1_rank0and1.png  image_1_rank2and3.png  image_2_rank0and1.png  image_2_rank2and3.png  image_3_rank0and1.png  image_3_rank2and3.png  image_4_rank0and1.png  image_4_rank2and3.png
```

File: optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py (hunk 3 above)
Comment (dsocek) [reply optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py:412]: Updated the code, now `use_distributed_cfg` is a bool argument (None is not a valid value)

File: optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py (hunk 4 above)
Comment (dsocek) [reply optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py]: This is __call__ section of the pipeline which does forward call / inference function of the pipeline with all its components.  We do parallelization at component level (running 1 call of transformer for noise prediction), so I think the location is OK.

File: optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py (hunk 3 above)
Comment (dsocek) [reply optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py:412]: Updated the code, now `use_distributed_cfg` is a bool argument (None is not a valid value)

File: optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py [hunk 5]
```diff
@@ -503,7 +525,8 @@ def __call__(
                 ceil(prompt_embeds.shape[1] / kernel_input_alignment_size) * kernel_input_alignment_size
             ) - prompt_embeds.shape[1]
             prompt_embeds = torch.nn.functional.pad(prompt_embeds, (0, 0, 0, pad_size))
-            negative_prompt_embeds = torch.nn.functional.pad(negative_prompt_embeds, (0, 0, 0, pad_size))
+            if negative_prompt_embeds is not None:
+                negative_prompt_embeds = torch.nn.functional.pad(negative_prompt_embeds, (0, 0, 0, pad_size))
```
Comment (yafshar) [inline optimum/habana/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py:529]:
```suggestion
            if pad_size > 0:
                prompt_embeds = torch.nn.functional.pad(prompt_embeds, (0, 0, 0, pad_size))
                if negative_prompt_embeds is not None:
                    negative_prompt_embeds = torch.nn.functional.pad(negative_prompt_embeds, (0, 0, 0, pad_size))
```

minor personal taste!

File: examples/stable-diffusion/README.md [hunk 6]
```diff
@@ -114,6 +114,34 @@ SD3 in quantization mode by setting runtime variable `QUANT_CONFIG=quantization/
 
 To run Stable Diffusion 3.5 Large, use `--model_name_or_path stabilityai/stable-diffusion-3.5-large` in the input.
 
+#### SD3 Distributed CFG Inference
+
+SD3 family of models use classifier-free guidance (CFG), which processes both conditional and unconditional latents during denoising, typically in
+a single forward pass with batch size 2. With the `--use_distributed_cfg` option, we parallelize this step across two HPU devices by splitting the
+batch and running each branch independently, then synchronizing to apply guidance. While this mode uses 2 HPUs per unique generated image, it
+achieves roughly 2x faster inference. Here is an example of running SD3.5-Large model with 2 HPU devices in disributed CFG mode:
```
Comment (yafshar) [inline examples/stable-diffusion/README.md:122]:
```suggestion
The SD3 family of models employs classifier-free guidance (CFG), which involves processing both conditional and unconditional latent
during the denoising phase, typically in a single forward pass using a batch size of 2. When the `--use_distributed_cfg` option is
enabled, this step is parallelized across two HPU devices by splitting the batch. Each device independently processes one branch
(conditional or unconditional), followed by synchronization to apply the guidance. Although this mode uses 2 HPUs per generated
image, it delivers approximately 2× faster inference. Below is an example of how to run the SD3.5-Large model using distributed
CFG mode with two HPU devices:
```

minor rewording, discard if you do not like it

**Review by yafshar — APPROVED** [review]
LGTM!

@regisss would you check this PR. It is ready for the final review!
